perun/collect/trace/run.py

```diff
@@ -1,206 +1,275 @@
-            UnicodeError, exceptions.InvalidBinaryException) as exception:
-        log.failed()
-        return _COLLECTOR_STATUS[systemtap.Status.EXCEPT][0], str(exception), dict(kwargs)
-
-
-def collect(**kwargs):
-    """ Runs the created SystemTap script and the profiled command
-
-    The output dictionary is updated with:
-     - output: path to the collector output file
-
-    :param dict kwargs: dictionary containing the configuration settings for the collector
-    :returns: (int as a status code, nonzero values for errors,
-              string as a status message, mainly for error states,
-              dict of kwargs and new values)
-    """
-    log.cprintln('Running the collector, progress output stored in collect_log_{0}.txt\n'
-                 'This may take a while... '.format(kwargs['timestamp']), 'white')
-    try:
-        # Call the system tap
-        code, kwargs['output_path'] = systemtap.systemtap_collect(**kwargs)
-        result = _COLLECTOR_STATUS[code]
-        if code != systemtap.Status.OK:
-            log.failed()
-            # Add specific return for STAP error which includes the precise log file path
-            if code == systemtap.Status.STAP:
-                return result[0], result[1].format(kwargs['log_path']), dict(kwargs)
-        return result[0], result[1], dict(kwargs)
-
-    except (OSError, CalledProcessError) as exception:
-        log.failed()
-        return CollectStatus.ERROR, str(exception), dict(kwargs)
-
-
-def after(**kwargs):
-    """ Handles the trace collector output and transforms it into resources
-
-    The output dictionary is updated with:
-     - profile: the performance profile contents created from the collector output
+    WD.header('Collect phase...')
 
-    :param kwargs: dictionary containing the configuration settings for the collector
-    :returns: tuple (int as a status code, nonzero values for errors,
+    # Check all the required dependencies
+    _check_dependencies()
+    # Try to lock the binary so that no other concurrent trace collector process can
+    # profile the same binary and produce corrupted performance data
+    ResourceLock(
+        LockType.Binary, os.path.basename(executable.cmd), kwargs['pid'], kwargs['locks_dir']
+    ).lock(kwargs['res'])
+    # Create the collect files
+    _create_collect_files(**kwargs)
+
+    # Extract and / or post-process the collect configuration
+    kwargs = strategy.extract_configuration(**kwargs)
+    if not kwargs['func'] and not kwargs['static'] and not kwargs['dynamic']:
+        msg = ('No profiling probes created (due to invalid specification, failed extraction or '
+               'filtering)')
+        return CollectStatus.ERROR, msg, dict(kwargs)
+
+    # Assemble script according to the parameters
+    stap_script.assemble_system_tap_script(kwargs['res'][Res.script()], **kwargs)
+
+    # Run the SystemTap and profiled command
+    systemtap.systemtap_collect(executable, **kwargs)
+
+    stdout.done('\n\n')
```

Comment:
Suggestion: Consider using perun/util/log functions. No output should go around the log if possible.

Reply:
Note that this is meant mostly for the Watchdog, now, I realized that stdout = log :).

Reply:
The default perun logging mechanisms for important events is used when watchdog is disabled - otherwise a separate log is created since most of the debug records are specialized to tracer and would needlessly clutter up the main perun log. At least that was the reasoning behind the decision.